File: 2023_algorithms/Advanced-Algorithms_2023-06-12/go_impls/fast_and_slow_pointers.go
package main

import (
	// "fmt"
	. "golang.org/x/exp/constraints"
)

type LinkedList[T Ordered] struct {
	Header *Node[T]
	Size   uint
}

type Node[T Ordered] struct {
	Value T
	Next  *Node[T]
}

func NewNode[T Ordered](value T) *Node[T] {
	return &Node[T]{Value: value, Next: nil}
}

func NewLinkedList[T Ordered]() *LinkedList[T] {
	return &LinkedList[T]{Header: nil, Size: 0}
}

func (list *LinkedList[T]) Push(value T) *LinkedList[T] {
	node := list.Header

	if node == nil {
		list.Size += 1
		list.Header = NewNode[T](value)
		return list
	}

	for node != nil {
		if node.Value == value {
			return list
		}

		if node.Next == nil {
			node.Next = NewNode[T](value)
			break
		}

		node = node.Next
	}

	list.Size += 1
	return list
}

func (list *LinkedList[T]) Get(idx int) *Node[T] {
	node := list.Header

	for i := 0; ; i++ {
		if i == idx {
			return node
		}
		node = node.Next
	}
}

func (list *LinkedList[T]) HasCycle() (*Node[T], bool) {
	slow, fast := list.Header, list.Header

	for {
		if fast == nil {
			return nil, false
		}

		slow = slow.Next
		if fast = fast.Next; fast == nil {
			return nil, false
		}
		fast = fast.Next

		if fast.Value == slow.Value {
			return fast, true
		}
	}
}

func (list *LinkedList[T]) CycleStart() *Node[T] {
	fast, hasCycle := list.HasCycle()
	if !hasCycle {
		return nil
	}

	slow := list.Header

	for {
		if fast.Value == slow.Value {
			return slow
		}

		slow, fast = slow.Next, fast.Next
	}
}

File: 2023_algorithms/Advanced-Algorithms_2023-06-12/src/fast_and_slow_pointers.rs
use super::linked_list_unique::{LinkedList, Next};
use std::fmt::Debug;

pub fn has_cycle<T: Debug + Copy + PartialEq>(list: &LinkedList<T>) -> (Next<T>, bool) {
    let mut slow = list.header.clone();
    let mut fast = list.header.clone();

    while fast.is_some() {
        fast = LinkedList::walk(fast, 2);
        slow = LinkedList::walk(slow, 1);

        /*
        let v = match fast {
            None => return (None, false),
            Some(ref v) => v,
        };

        if v.borrow().value == slow.clone().unwrap().borrow().value {
            return (fast, true);
        }
        */

        let (v1, v2) = match (&slow, &fast) {
            (Some(v1), Some(v2)) => (v1, v2),
            _ => return (None, false),
        };

        if v1.borrow().value == v2.borrow().value {
            return (fast, true);
        }
    }

    (None, false)
}

pub fn cycle_start<T: Debug + Copy + PartialEq>(list: &LinkedList<T>) -> Next<T> {
    let (mut fast, has_cycle) = has_cycle(list);
    if !has_cycle {
        return None;
    }

    let mut slow = list.header.clone();

    loop {
        let (v1, v2) = match (&slow, &fast) {
            (Some(v1), Some(v2)) => (v1, v2),
            _ => return None,
        };

        if v1.borrow().value == v2.borrow().value {
            return slow;
        }

        slow = LinkedList::walk(slow, 1);
        fast = LinkedList::walk(fast, 1);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn t_cycle() {
        let mut list = LinkedList::new();
        list.push(1).push(2).push(3).push(2).push(4).push(5);

        let n2 = list.get(1);
        let last = list.last();
        assert_eq!(LinkedList::walk(n2.clone(), 2).unwrap().borrow().value, 4);

        last.unwrap().borrow_mut().next = n2;
        let (next, ok) = has_cycle(&list);
        assert!(ok);
        assert_eq!(next.unwrap().borrow().value, 5);

        assert_eq!(cycle_start(&list).unwrap().borrow().value, 2);
    }
}
